### PR TITLE
Engine metrics module implementation

### LINKED ISSUE #16031: Implement the metrics module and integrate it with other modules
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 5.0 | Engine | Manager| - | - |

## Description
The objective of the this task is Implement the structure of the metrics module, and integrate it with other modules, in the engine. 

## DoD
- [x] Implement first iteration of the metrics module.
  - [x] Implement the most convenient method according to what was analyzed in issue #15993, it can be a dependency injection or it can work as a singleton with enabling macros.
  - [x] Define a clear interface in order to be able to implement the indicators in the various modules.
    - Commands:
      - [x] dump
      - [x] enable `<metricName>` [true | false]  - allow to enable or disable acquisition metric.
      - [x] list - print registered metrics with following format, name: `"cont1" status: "enable|disable" type: "counter" value: "12"`
      - [x] get `<metricName>` - print a single metric as pretty json.
  - [x] Able to Collect at least a signal from any module.
- [x] Update Documentation.

### PR DESCRIPTION
**Pull request by NahuFigueroa97**
|Related issue|
|---|
|closes #16031 |

## Description
All the components of the system should be observable from outside, reporting metrics about what they are doing, and how quickly they are doing it. The objective of the this PR is Implement the structure of the metrics module in the engine.

## Configuration options

The configuration file to describe the type of each metrics.
path `/var/ossec/engine/store/metrics/config/0`

``` json
[
    {
        "name": "CountExample",
        "dataHubEnable" : true,
        "enable": true,
        "signalType": "meter",
        "instrumentType" : "counter",
        "subType": "uint64",
        "outputFile": "countExample.txt",
        "exportIntervalMillis": 100,
        "exportTimeoutMillis": 50,
        "unit": "some unit",
        "description": "some description"
    },
    {
        "name": "HistogramExample",
        "dataHubEnable" : false,
        "enable": true,
        "signalType": "meter",
        "instrumentType" : "histogram",
        "subType": "double",
        "outputFile": "histogramExample.txt",
        "exportIntervalMillis": 100,
        "exportTimeoutMillis": 50,
        "unit": "some unit",
        "description": "some description"
    },
    {
        "name": "UpDownCountExample",
        "dataHubEnable" : false,
        "enable": true,
        "signalType": "meter",
        "instrumentType" : "upDownCounter",
        "subType": "int64",
        "outputFile": "upDownCounterExample.txt",
        "exportIntervalMillis": 100,
        "exportTimeoutMillis": 50,
        "unit": "some unit",
        "description": "some description"
    },
    {
        "name": "ObservableGaugeExample",
        "dataHubEnable" : false,
        "enable": true,
        "signalType": "meter",
        "instrumentType" : "observableGauge",
        "subType": "double",
        "outputFile": "observableGauge.txt",
        "exportIntervalMillis": 100,
        "exportTimeoutMillis": 50,
        "unit": "some unit",
        "description": "some description"
    },
    {
        "name": "TracerExample",
        "dataHubEnable" : false,
        "enable": true,
        "signalType": "tracer",
        "exporterType": "logging",
        "outputFile": "tracerExample.txt",
        "processorType": "simple",
        "description": "some description"
    }
]
```

### :green_circle: Dump command

```
/var/ossec/engine/wazuh-engine metrics dump
[{"scope":"CountExample","schema":"","version":"","records":[{"start_time":"Tue Mar  7 14:46:09 2023","instrument_name":"CountExample_counter","instrument_description":"some description","unit":"some unit","attributes":[{"type":"SumPointData","value":1}]}]}]

```
### :green_circle: Enable command

- With this command we disable some instrument by its name
```
/var/ossec/engine/wazuh-engine metrics enable "CountExample" false
```

- if the instrument name does not exist it throws an error:
```
/var/ossec/engine/wazuh-engine metrics enable "CountExampleNotExist" true
The Instrument CountExampleNotExist has not been created.

```
### :green_circle: Get command

- With this command we get some instrument by name
```
/var/ossec/engine/wazuh-engine metrics get "CountExample"
```

- if the instrument name does not exist it throws an error:
```
/var/ossec/engine/wazuh-engine metrics get "CountExampleNotExist"
Instrument 'CountExampleNotExist' does not exist or wasn't loaded

```

### :green_circle: List command

- With this command we show the name, status and type of all configured instruments.
```
/var/ossec/engine/wazuh-engine metrics list
```

### :green_circle: Test command

- With this command we generate dummy metrics for testing.
```
/var/ossec/engine/wazuh-engine metrics test
```

### CODE REVIEW

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * @brief Get the Metrics's name from the params or return an error
+ *
+ * @param params The json /data from the request
+ * @return [bool, std::string] True if the name is valid, false otherwise
+ *                             The name if it's valid, the error message otherwise.
+ */
+std::tuple<bool, std::string> getNameOrError(const json::Json& params);
+api::CommandFn metricsDumpCmd();
```
Comment (soynof) [inline src/engine/source/api/include/api/metrics/commands.hpp:21]: Add function descripción

Comment (NahuFigueroa97) [inline src/engine/source/api/include/api/metrics/commands.hpp:21]: Severe - Documentation : add documentation to methods

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+api::CommandFn metricsGetCmd();
```
Comment (soynof) [inline src/engine/source/api/include/api/metrics/commands.hpp:22]: Add function descripción

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+api::CommandFn metricsEnableCmd();
```
Comment (soynof) [inline src/engine/source/api/include/api/metrics/commands.hpp:23]: Add function descripción

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+api::CommandFn metricsListCmd();
```
Comment (soynof) [inline src/engine/source/api/include/api/metrics/commands.hpp:24]: Add function descripción

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+api::CommandFn metricsTestCmd();
```
Comment (soynof) [inline src/engine/source/api/include/api/metrics/commands.hpp:25]: Add function descripción

File: src/engine/source/api/include/api/metrics/commands.hpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+void registerAllCmds(std::shared_ptr<api::Registry> registry);
```
Comment (JorgeMarinoDev) [inline src/engine/source/api/include/api/metrics/commands.hpp:27]: Add missing documentation for Cmd()s

File: src/engine/source/metrics/include/metrics.hpp [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+     */
+    std::shared_ptr<DataHub> getDataHub();
+    
+    /**
+     * @brief Set the next handle on the chain
+     *
+     * @param handler handler that will be set as the next on the chain
+     * @return std::shared_ptr<Handler> the same handler that has been set as
+     * next
+     */
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/metrics.hpp:54]: severe - documentation: Fix documentation in all methods

File: src/engine/source/api/src/metrics/commands.cpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            std::get<json::Json>(result),
+            fmt::format("Metrics successfully geted")};
+    };
+}
+
+api::CommandFn metricsEnableCmd()
+{
+    return [](const json::Json& params) -> api::WazuhResponse
+    {
+        auto name = params.getString("/nameInstrument");
```
Comment (soynof) [inline src/engine/source/api/src/metrics/commands.cpp:68]: Rename "/nameInstrument" to "/instrumentName"

File: src/engine/source/metrics/include/chainOfResponsibility.hpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        return data;
+    }
+
+    /**
+     * @brief Get next step on the chain
+     *
+     * @return std::shared_ptr<AbstractHandler> next handler on the chain
+     */
+    std::shared_ptr<AbstractHandler> next() const
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/chainOfResponsibility.hpp:86]: improve - desing: this method is unsed

File: src/engine/source/cmds/include/cmds/metrics.hpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+namespace details
+{
+constexpr auto ORIGIN_NAME = "engine_integrated_metrics_api";
+constexpr auto API_METRICS_DUMP_SUBCOMMAND {"dump"};
+constexpr auto API_METRICS_GET_SUBCOMMAND {"get"};
+constexpr auto API_METRICS_ENABLE_SUBCOMMAND {"enable"};
+constexpr auto API_METRICS_LIST_SUBCOMMAND {"list"};
+constexpr auto API_METRICS_TEST_SUBCOMMAND {"test"};
+
+std::string commandName(const std::string& command);
```
Comment (soynof) [inline src/engine/source/cmds/include/cmds/metrics.hpp:25]: Add function description

File: src/engine/source/metrics/include/dataHubExporter.hpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    void
+    printAttributes(const std::map<std::string, sdk::common::OwnedAttributeValue>& map,
+                    const std::string prefix);
+    void printResources(json::Json& jsonObj, const opentelemetry::sdk::resource::Resource& resources);
+};
+} // namespace metrics
+} // namespace exporter
+OPENTELEMETRY_END_NAMESPACE
+
+#endif // __JSON_EXPORTER_H
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/dataHubExporter.hpp:87]: Add new line

File: src/engine/source/api/src/metrics/commands.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+        return api::WazuhResponse("OK");
+    };
+}
+
+api::CommandFn metricsListCmd()
+{
+    return [](const json::Json& params) -> api::WazuhResponse
+    {
+        auto result = Metrics::instance().getListInstruments();
```
Comment (JorgeMarinoDev) [inline src/engine/source/api/src/metrics/commands.cpp:87]: typo:   getListInstruments  -> getInstrumentsList

File: src/engine/source/metrics/include/exporterHandler.hpp [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+#define _EXPORTER_HANDLER_HPP
+
+#include "chainOfResponsibility.hpp"
+#include "metricsContext.hpp"
+
+class ExporterHandler final : public AbstractHandler<std::shared_ptr<MetricsContext>>
+{
+public:
+    /**
+     * @brief Trigger for the normalization process on the metrics context.
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/exporterHandler.hpp:11]: improve - documentation: this isn't a normalizacion process

File: src/engine/source/metrics/include/exporterHandler.hpp [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+     *
+     * @param data
+     * @return std::shared_ptr<MetricsContext>
+     */
+    virtual std::shared_ptr<MetricsContext> handleRequest(std::shared_ptr<MetricsContext> data);
+
+private:
+    /**
+     * @brief
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/exporterHandler.hpp:20]: improve - documentation: add documentation

File: src/engine/source/metrics/include/providerHandler.hpp [hunk 15]
```diff
@@ -0,0 +1,27 @@
+#ifndef _PROVIDER_HANDLER_HPP
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/providerHandler.hpp:1]: improve - documentation: same to exporter

File: src/engine/source/metrics/include/readerHandler.hpp [hunk 16]
```diff
@@ -0,0 +1,27 @@
+#ifndef _READER_HANDLER_HPP
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/readerHandler.hpp:1]: improve - documentation: same to exporter

File: src/engine/source/metrics/include/metricsContext.hpp [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+enum class ProcessorsTypes
+{
+    Batch,
+    MultiProcessor,
+    Simple,
+};
+
+enum class ProviderTypes
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/metricsContext.hpp:31]: improve -  naming: it is SignalsTypes

File: src/engine/source/metrics/include/processorHandler.hpp [hunk 18]
```diff
@@ -0,0 +1,27 @@
+#ifndef _PROCESSOR_HANDLER_HPP
```
Comment (NahuFigueroa97) [inline src/engine/source/metrics/include/processorHandler.hpp:1]: improve - documentation: same to exporter

File: src/engine/source/cmds/src/catalog.cpp [hunk 19]
```diff
@@ -14,6 +14,7 @@
 
 #include "apiclnt/client.hpp"
 #include "defaultSettings.hpp"
+#include "metrics.hpp"
```
Comment (NahuFigueroa97) [inline src/engine/source/cmds/src/catalog.cpp:17]: delete this

Comment (cborla) [inline src/engine/source/cmds/src/catalog.cpp:17]: I believe that the library should not be included here.

File: src/engine/source/cmds/src/start.cpp [hunk 20]
```diff
@@ -222,6 +224,13 @@ void runStart(ConfHandler confManager)
             router->addRoute(routeName, routePriority, routeFilter, routeEnvironment);
         }
 
+        // Initialize Metrics Module
+        Metrics::instance().initMetrics("engine-metrics", "/var/ossec/engine/store/metrics/config/0");
```
Comment (JorgeMarinoDev) [inline src/engine/source/cmds/src/start.cpp:228]: Add TODO: Resolve fullpath accordingly.

File: src/engine/source/metrics/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+####################################################################################################
+# Dependencies
+####################################################################################################
+CPMAddPackage(
+    NAME opentelemetry
+    GITHUB_REPOSITORY open-telemetry/opentelemetry-cpp
+    GIT_TAG v1.8.2
+    VERSION 1.8.2
+    EXCLUDE_FROM_ALL YES
+#    DOWNLOAD_ONLY
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/CMakeLists.txt:14]: Remove Commented line.

File: src/engine/source/metrics/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    OPTIONS
+        "BUILD_TESTING OFF"
+        "WITH_ZIPKIN ON"
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/CMakeLists.txt:17]: Add TODO: Add build option and link zipkin to it. Define OT modules policy.

File: src/engine/source/metrics/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+opentelemetry_exporter_ostream_span
+opentelemetry_exporter_in_memory
+opentelemetry_exporter_zipkin_trace
+opentelemetry_exporter_ostream_metrics
+opentelemetry_resources
+PRIVATE base)
+
+target_include_directories(metrics PUBLIC
+${ENGINE_METRICS_INCLUDE_DIR}
+${FETCHCONTENT_BASE_DIR}/opentelemetry-src/exporters/zipkin/include
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/CMakeLists.txt:47]: same here . zipkin,

File: src/engine/source/metrics/include/dataHubExporter.hpp (hunk 11 above)
Comment (JorgeMarinoDev) [reply src/engine/source/metrics/include/dataHubExporter.hpp:87]: +1

File: src/engine/source/metrics/include/dataHubInterface.hpp [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+class DataHubInterface
+{
+public:
+    /// @brief updates the data of the referenced object
+    /// @param scope name of the resource scope
+    /// @param object json object with updated information
+    virtual void setResource(const std::string& scope, json::Json object) = 0;
+};
+
+#endif // _DATA_HUB_INTERFACE_H
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/include/dataHubInterface.hpp:16]: add new line

File: src/engine/source/metrics/include/metrics.hpp [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::Counter<double>>> m_doubleCounter;
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::Counter<uint64_t>>> m_uint64Counter;
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::Histogram<double>>> m_doubleHistogram;
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::Histogram<uint64_t>>> m_uint64Histogram;
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::UpDownCounter<double>>> m_doubleUpDownCounter;
+    std::unordered_map<std::string, opentelemetry::nostd::unique_ptr<opentelemetry::metrics::UpDownCounter<int64_t>>> m_int64UpDownCounter;
+    std::unordered_map<std::string, opentelemetry::nostd::shared_ptr<opentelemetry::metrics::ObservableInstrument>> m_doubleObservableGauge;
+    std::unordered_map<std::string, opentelemetry::nostd::shared_ptr<opentelemetry::metrics::ObservableInstrument>> m_int64ObservableGauge;
+    opentelemetry::context::Context m_context;
+    std::unordered_map<std::string, bool> controller;
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/include/metrics.hpp:275]: m_controller . severe. set descriptive name to the map.

File: src/engine/source/metrics/include/metricsContext.hpp [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+    std::unique_ptr<opentelemetry::sdk::trace::SpanProcessor> processor;
+    std::shared_ptr<opentelemetry::trace::TracerProvider> traceProvider;
+    std::shared_ptr<opentelemetry::sdk::metrics::MeterProvider> meterProvider;
+    std::shared_ptr<opentelemetry::exporter::memory::InMemorySpanData> inMemorySpanData;
+    std::ofstream file;
+    int timeIntervalBetweenExports;
+    int numSpans;
+
+    std::chrono::milliseconds export_interval_millis;
+    std::chrono::milliseconds export_timeout_millis;
```
Comment (JorgeMarinoDev) [inline src/engine/source/metrics/include/metricsContext.hpp:81]: add TODO: Write proper doc to these members.

**Review by JorgeMarinoDev — CHANGES_REQUESTED** [review]

**Comment by JorgeMarinoDev** [comment]
Also please sign the unsigned commits. @NahuFigueroa97 
https://github.com/wazuh/wazuh/pull/16339/commits/e8c22e0177992d2db5f93858c78ec02564b8c7e6
https://github.com/wazuh/wazuh/pull/16339/commits/22c178d5fb81f8f3be8162594ead907a2ad2d351

**Review by cborla — APPROVED** [review]
LGTM!

**Review by soynof — APPROVED** [review]

**Review by JorgeMarinoDev — APPROVED** [review]
LGTM

**Pull request merged**